---
title: "PanMyeloid Assignment--Data Exploration"
author: "ZHANG, Yiling"
date: "`r Sys.Date()`"
output:
  rmdformats::robobook:
    code_folding: show
    self_contained: true
    thumbnails: false
    lightbox: true
    df_print: paged
    gallery: false
    highlight: tango
pkgdown:
  as_is: true    
---


```{r include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.align = "center",
  fig.width = 6,
  autodep = TRUE,
  cache.lazy = FALSE
)
```

# 1. Environment Setting

```{r}
# Single-cell required
library(Seurat)
library(SeuratData)
library(SeuratDisk)
library(SeuratWrappers)

# For data analysis
library(dplyr)
library(tidyr)
library(stringr)

# For plotting
library(ggplot2)
library(cowplot)
library(ggridges)
library(ggsci)
```

```{r}
packageVersion("Seurat")
```

# 2. Import and Examine Data

```{r}
hfile <- Connect("../processedData/alldata.obj/regressed-alldata.h5seurat")
hfile$index()
```

```{r include=FALSE}
hfile$close_all()
```

```{r cache=TRUE}
# Convert("../processedData/alldata.obj/regressed-alldata.h5ad", dest = "h5seurat", overwrite = TRUE)
alldata <- LoadH5Seurat("../processedData/alldata.obj/regressed-alldata.h5seurat",meta.data=F)
alldata.meta <- read.csv("../processedData/alldata.obj/alldata-metadata.csv",header = T, row.names = 1)
alldata@meta.data <- alldata.meta
alldata
```

```{r}
head(alldata@meta.data)
```

```{r}
alldata@meta.data %>% 
    ggplot(aes(x=patient, fill=patient)) + 
    geom_bar() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold"),legend.position="none") +
    ggtitle("NCells")
```
* Cell number check
```{r fig.width=12}
cell_tech <- alldata@meta.data %>% 
  ggplot(aes(fill=tech_10X, x=tech_10X)) + 
  geom_bar(alpha = 0.75) + ggtitle("NCells") +
  theme_classic() +
  scale_fill_npg()
cell_cancer <- alldata@meta.data %>% 
  ggplot(aes(fill=cancer, x=cancer)) + 
  geom_bar(alpha = 0.75) + ggtitle("NCells") +
  theme_classic() +
  scale_fill_simpsons()
plot_grid(cell_tech,cell_cancer,rel_widths = c(1,2))
```

* UMI counts (transcripts) per cell
```{r fig.width=12}
count_tech <- alldata@meta.data %>% 
  ggplot(aes(fill=tech_10X, x=n_counts)) + 
  geom_density(alpha = 0.3) + 
  scale_x_log10() + 
  theme_classic() +
  scale_fill_npg()
count_cancer <- alldata@meta.data %>% 
  ggplot(aes(fill=cancer, x=n_counts)) + 
  geom_density(alpha = 0.3) + 
  scale_x_log10() + 
  theme_classic() +
  scale_fill_simpsons()
plot_grid(count_tech,count_cancer,rel_widths = c(2,2))
```

# 3. Dimensionality Reduction

```{r cache = TRUE,fig.width = 5}
alldata <- FindVariableFeatures(alldata,selection.method = "vst",nfeatures = 2000)
alldata <- RunPCA(alldata, npcs = 100, verbose = FALSE)
ElbowPlot(alldata,ndims = 100)
```

```{r}
# Set dim
pc.num=1:100
```

```{r cache=TRUE}
# Run UMAP
alldata <- RunUMAP(alldata, dims = pc.num, reduction = "pca")
alldata <- RunTSNE(alldata, dims = pc.num)
```

```{r fig.width = 12}
p1 <- DimPlot(alldata, group.by = "patient", reduction = "umap") + NoLegend()
p2 <- DimPlot(alldata, group.by = "patient", reduction = "tsne") + NoLegend()
plot_grid(p1,p2,rel_widths = c(2,2))
```

```{r fig.width = 12}
p3 <- DimPlot(alldata, group.by = "tech_10X", reduction = "umap") 
p4 <- DimPlot(alldata, group.by = "tech_10X", reduction = "tsne")
plot_grid(p3,p4,rel_widths = c(2,2))
```

```{r fig.width = 12}
p5 <- DimPlot(alldata, group.by = "tech", reduction = "umap") 
p6 <- DimPlot(alldata, group.by = "tech", reduction = "tsne")
plot_grid(p5,p6,rel_widths = c(2,2))
```

```{r fig.width = 12}
p7 <- DimPlot(alldata, group.by = "MajorCluster", reduction = "umap") 
p8 <- DimPlot(alldata, group.by = "MajorCluster", reduction = "tsne")
plot_grid(p7,p8,rel_widths = c(2,2))
```

# 4. Harmony Integration

```{r}
library(harmony)
library(Rcpp)
library(rlang)
```

```{r}
dat.h <- alldata
```

```{r}
timestart<-Sys.time()

integrated.by.both.h <- harmony::RunHarmony(dat.h, group.by.vars =c("tech_10X","patient"), assay.use = "RNA", plot_convergence = TRUE)

timeend<-Sys.time()
```

```{r}
runningtime<-timeend-timestart
print(runningtime)
```

```{r}
# Run UMAP
integrated.by.both.h <- RunUMAP(integrated.by.both.h, dims = 1:100, reduction = "harmony")
```

```{r fig.width = 12}
p_tech <- DimPlot(object = integrated.by.both.h, group.by = "tech_10X", reduction = "umap")
p_patient <- DimPlot(object = integrated.by.both.h, group.by = "patient", reduction = "umap") +NoLegend()
plot_grid(p_tech,p_patient,rel_widths = c(2,2))
```

# 4. LIGER Integration
```{r}
library(rliger)
```

```{r}
dat.l <- alldata
dat.l
```

```{r}
dat.l <- ScaleData(dat.l, split.by = "patient", do.center = FALSE)
dat.l <- RunOptimizeALS(dat.l, k = 20, lambda = 5, split.by = "patient")
dat.l <- RunQuantileNorm(dat.l, split.by = "patient")
```

# 5. Seurat3 Integration

---

Posing error: caught segfault, address (nil), cause 'memory not mapped'  

[Issue#](https://github.com/satijalab/seurat/issues/4628) exists on GitHub with bug label but no responses.  

Maybe due to insufficient menory? Anyway, just show the codes below.

---

```{r eval=FALSE}
dat.s <- alldata
```

```{r eval=FALSE}
split_seurat <- SplitObject(dat.s, split.by = "patient")

# Normalize and identify variable features for each dataset independently
split_seurat <- lapply(X = split_seurat, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration
integ_features <- SelectIntegrationFeatures(object.list = split_seurat)
```

Perform CCA, find the best buddies or anchors and filter incorrect anchors.  
```{r eval=FALSE}
# Find best buddies - can take a while to run
integ_anchors <- FindIntegrationAnchors(object.list = split_seurat, 
                                        normalization.method = "LogNormalize", 
                                        anchor.features = integ_features,
                                        dims = 1:20, reduction = "rpca")
```

Integrate across conditions.
```{r eval=FALSE}
# Integrate across conditions
integrated.s <- IntegrateData(anchorset = integ_anchors, normalization.method = "LogNormalize")
```


# 5. LISI Evaluation

```{r}
# devtools::install_github("immunogenomics/lisi")
library(lisi)
library(Rcpp)
```

## 5.1 Check Before Integration
```{r}
umap.co <- Embeddings(object =alldata[["umap"]])

meta.df <- data.frame(patient = alldata$patient, tech = alldata$tech, tech_10X = alldata$tech_10X)
```

```{r eval=FALSE}
umap.res <- compute_lisi(umap.co, meta.df, c("patient", "tech", "tech_10X"))
# save(umap.res,tsne.res,file = "../processedData/tmp.data/ori-lisi-res.RData")
```

```{r}
summary(umap.res)
```


> Each row in the output data frame corresponds to a cell from X. The score (e.g. max 13.282) indicates the effective number of different categories represented in the local neighborhood of each cell. If the cells are well-mixed, then we might expect the LISI score to be near 46 for a categorical variable with 46 categories(e.g. "patient").

## 5.2 Harmony Output

```{r}
h.both.umap.co <- Embeddings(object =integrated.by.both.h[["umap"]])
h.both.umap.res <- compute_lisi(h.both.umap.co, meta.df, c("patient", "tech", "tech_10X"))
```

```{r}
summary(h.both.umap.res)
```


# Supplemental Information

<details><summary>**▶ Session Info**</summary>
```{r echo=FALSE}
sessionInfo()
```
</details>


<details><summary>**▶ Server Info**</summary>
* Linux Version
```{bash echo=FALSE}
cat /etc/os-release
```

* CPU info
```{bash echo=FALSE}
lscpu | head -25
```
</details>